library/x509.c

```diff
@@ -1019,7 +1019,8 @@ int mbedtls_x509_self_test( int verbose )
         if( verbose != 0 )
             mbedtls_printf( "failed\n" );
 
-        return( ret );
+        mbedtls_x509_crt_free( &clicert );
```

Comment:
It is more in line with the remainder of the library to do a bulk initialization at the beginning. I'd therefore suggest to move the initialization call `mbedtls_x509_crt_init( &cacert );` to L1013.